rosbag2_cpp/src/rosbag2_cpp/reindexer.cpp

```diff
@@ -0,0 +1,252 @@
+  u_int32_t second_db_num = std::stoul(second_match.str(1), nullptr, 10);
+
+  return first_db_num < second_db_num;
+}
+
+std::vector<rcpputils::fs::path> Reindexer::get_bag_files(
+  const rcpputils::fs::path & base_folder)
+{
+  // Look in the uri directory to see what database files are there
+  std::vector<rcpputils::fs::path> output;
+
+  auto allocator = rcutils_get_default_allocator();
+  auto dir_iter = rcutils_dir_iter_start(base_folder.string().c_str(), allocator);
+
+  // Make sure there are files in the directory
+  if (dir_iter == nullptr) {
+    throw std::runtime_error("Empty directory.");
+  }
+
+  // Get all file names in directory
+  do {
+    auto found_file = rcpputils::fs::path(dir_iter->entry_name);
+
+    // We are ONLY interested in database files
+    if (found_file.extension().string() == ".db3") {
+      auto full_path = base_folder / found_file;
+      output.emplace_back(full_path);
+    }
+  } while (rcutils_dir_iter_next(dir_iter));
+
+  // Sort relative file path by database number
+  std::sort(
+    output.begin(), output.end(),
+    [](rcpputils::fs::path a, rcpputils::fs::path b) {return compare_relative_file(a, b);});
+
+  return output;
+}
+
+void Reindexer::fill_topics_metadata()
+{
+  // rcpputils::check_true(storage_ != nullptr, "Bag is not open. Call open() before reading.");
+  topics_metadata_.clear();
+  topics_metadata_.reserve(metadata_.topics_with_message_count.size());
+  for (const auto & topic_information : metadata_.topics_with_message_count) {
+    topics_metadata_.push_back(topic_information.topic_metadata);
+  }
+}
+
+void Reindexer::init_metadata(
+  const std::vector<rcpputils::fs::path> & files,
+  const rosbag2_storage::StorageOptions & storage_options)
+{
+  metadata_ = rosbag2_storage::BagMetadata{};
+
+  // This reindexer will only work on SQLite files, so this can't change
+  metadata_.storage_identifier = storage_options.storage_id;
+  metadata_.starting_time = std::chrono::time_point<std::chrono::high_resolution_clock>(
+    std::chrono::nanoseconds::max());
+
+  // Record the relative paths to the metadata
+  for (const auto & path : files) {
+    auto cleaned_path = path.filename().string();
+    metadata_.relative_file_paths.push_back(cleaned_path);
+  }
+}
+
+void Reindexer::aggregate_metadata(
+  const std::vector<rcpputils::fs::path> & files,
+  const rosbag2_storage::StorageOptions & storage_options)
+{
+  std::map<std::string, rosbag2_storage::TopicInformation> temp_topic_info;
+
+  // In order to most accurately reconstruct the metadata, we need to
+  // visit each of the contained relative database files in the bag,
+  // open them, slurp up the info, and stuff it into the master
+  // metadata object.
+  auto metadata_io_default = std::make_unique<rosbag2_storage::MetadataIo>();
+  auto bag_reader = std::make_unique<rosbag2_cpp::readers::SequentialReader>(
+    std::move(storage_factory_), converter_factory_, std::move(metadata_io_default));
```

Comment:
This move of the storage_factory is interesting. Ownership is never transferred back to the Reindexer - so if you call `reindex` twice, you will run into an error on the second call, meaning that the API as presented is broken.

This start to look to me like "should the reindexer even be a class? Or is it just a function?"

Reply:
I would not mind converting the Reindexer to a purely functional implementation. It's basically a one-and-done operation, so having an object made out of it could be considered a sort of "waste of space".

Would you like me to make the changes? If so, where should these functions live? A new namespace named something like `reindex`? Or just have them sort of float around in `rosbag2_cpp`?

Reply:
Up to you - either make sure the class API is consistent, or refactor into being a single public function `rosbag2_cpp::reindex_bag` (with private implementation in an unnamed namespace in the .cpp). Either solution works for me. The class makes the most sense if you forsee a need for reuse - e.g. do we ever want to call the same reindexer multiple times, or is it fine to just construct new reindexer for every reindexing operation (I'm leaning to the latter)

Reply:
When I was originally writing the reindexer, I had actually considered doing it all as one function. It's just that the other operations were all classes, so I wanted to fit in with how you guys were doing things. But I think that having the reindexer as just a function makes a lot of sense - it's only really getting run once per bag, and doesn't need to remember any information from run to run. There's no real "state", in that regard.

So I think that it's certainly doable. I'll finish up getting the storage API to be able to pass the expected file extension (it's looking like the storage is a protected member variable of the reader, so I may have to make a new  function for it; unless you know of a better way to query that info?), and then I'll get started on this.

Reply:
You have the `StorageFactory` - so you could construct your own storage instance just to call this function

Reply:
Unfortunately, I still run into the same problem: In order for the storage factory to create a storage instance, it needs a proper `storage_options` URI, complete with extension. Otherwise, I get errors like:
    Could not open `/rosbag2_cpp/test/rosbag2_cpp/reindex_test_bags/multiple_files` with `sqlite3`. Error: Failed to setup storage. Error: Could not read-only open database. SQLite error (10): disk I/O error

Reply:
Let's just go ahead and ignore the suffix for now then - I don't want to block on getting this 100% perfect - let's just avoid hardcoding `sqlite3` stuff